### PR TITLE
Fix: Ampersand in post tag causes editor crash

### PR DESCRIPTION
**Pull request by jorgefilipecosta**
## Description
Fixes: https://github.com/WordPress/gutenberg/issues/12832

This PR changes FlatTermSelector to always unescapes the terms names when they come from the server (the server escapes them), and always escapes when creating terms, or searching terms, to make sure they are searched in the same way they are stored on the server.

## How has this been tested?
Pasted "foo & bar" in the tag input field, pressed enter and repeated both steps again. (the editor should not crash)

I checked that the terms with special characters are saved in the database in the same way they are saved if the classic editor (in escaped form).

I checked that a similar problem was not happening on the hierarchical terms.

### CODE REVIEW

File: packages/editor/src/components/post-taxonomies/flat-term-selector.js [hunk 1]
```diff
@@ -39,6 +49,17 @@ class FlatTermSelector extends Component {
 		};
 	}
 
+	unescapeTerm( term ) {
```
Comment (aduth) [inline packages/editor/src/components/post-taxonomies/flat-term-selector.js]: Why is `unescapeTerm` deserving of a utility function, but the inverse `escapeTerm` is effectively performed inline in `findOrCreateTerm`. Should we be consistent? Can this function be documented?

Comment (aduth) [reply packages/editor/src/components/post-taxonomies/flat-term-selector.js]: Order-wise, I'd prefer if we keep a convention of `constructor`, lifecycle, utility, `render` in components (i.e. move these utility functions under `componentDidMount` and friends).

Maybe worth enforcing via lint rules?

Comment (aduth) [reply packages/editor/src/components/post-taxonomies/flat-term-selector.js]: I don't have a strong sense one way or the other, but it's unclear to me that this is something specific to the component, vs. a generalized function which could be defined outside the component definition (also allows for easier unit testing).

I might also be thinking this way in consideration of how we might want to have this data be managed in the future. Ideally we don't have to keep this bizarre escape dancing in this component itself, but rather allow it to be managed somewhere in `@wordpress/core-data`. At that point, we might still have `unescapeTerm`, but it certainly wouldn't be component-specific.

**Review by aduth — CHANGES_REQUESTED** [review]
While I can now add the term without a crash occurring, when the post is saved, the "Tags" field reverts back to an empty value. Are you able to reproduce?

1. New post
2. Add title
3. Add tag "foo & bar"
4. Click "Save Draft"
5. Observe, after save finishes, the tag has been removed, and is also not present upon reload

**Comment by jorgefilipecosta** [comment]
> New post
Add title
Add tag "foo & bar"
Click "Save Draft"
Observe, after save finishes, the tag has been removed, and is also not present upon reload

Hi @aduth thank you for the review. That issue was caused because I missed stash&committ a bind for the unescape functions. The problem should not happen in the new logic.

File: packages/editor/src/components/post-taxonomies/flat-term-selector.js (hunk 1 above)
Comment (jorgefilipecosta) [reply packages/editor/src/components/post-taxonomies/flat-term-selector.js]: > Why is unescapeTerm deserving of a utility function, but the inverse escapeTerm is effectively performed inline in findOrCreateTerm. Should we be consistent?

We don't have an escape function that escapes term objects; we escape the term name.  The function to escape the term name is just the function to escape strings in lodash. I thought about having escapeTermName and unescapeTermName functions, but they would be just an import as: `escape as escapeTermName,`and later if we need other escapes, it stops making sense.
Off course we can have escapeTerm escape the term and get the name after but that seems like an unnecessary waste of resources as we just need the escaped name and we will not save the escaped term anywhere.


> Order-wise, I'd prefer if we keep a convention of constructor, lifecycle, utility, render in components (i.e. move these utility functions under componentDidMount and friends).

> Maybe worth enforcing via lint rules?

Having a lint rule for that would be an excellent idea, it is a rule that we may easily miss.


> I don't have a strong sense one way or the other, but it's unclear to me that this is something specific to the component, vs. a generalized function which could be defined outside the component definition (also allows for easier unit testing).

I agree with this; the functions were moved outside the component and documentation was added to them.

**Review by aduth — APPROVED** [review]

**Comment by KokkieH** [comment]
Thank you for the fix!

**Pull request merged**